analyzers/src/SonarAnalyzer.Common/Rules/ExpressionComplexityBase.cs

```diff
@@ -43,26 +43,44 @@ public abstract class ExpressionComplexityBase<TExpression> : ParameterLoadingDi
             rule = Language.CreateDescriptor(DiagnosticId, MessageFormat, isEnabledByDefault: false);
 
         protected sealed override void Initialize(ParameterLoadingAnalysisContext context) =>
-            context.RegisterSyntaxTreeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
                 {
-                    var root = c.Tree.GetRoot();
-                    var rootExpressions = NoncompoundSubexpressions(root);
-                    var compoundExpressionsDescendants = root.DescendantNodes().Where(IsCompoundExpression).SelectMany(NoncompoundSubexpressions);
-
-                    foreach (var expression in rootExpressions.Concat(compoundExpressionsDescendants))
+                    if (NodeKind(c.Node.Parent) is TSyntaxKind parentKind && (ComplexityIncreasingKinds.Contains(parentKind) || TransparentKinds.Contains(parentKind)))
+                    {
+                        // The parent of the expression is itself complexity increasing (e.g. &&) or transparent (e.g. parenthesis).
```

Comment:
I don't see the benefit of this comment on line 50 - it's basically reading the code inside the `if` at line 48

Lines 51-52 are necessary because they explain _why_ it's necessary and give an example